### PR TITLE
refactor!: use pointers in struct fields

### PR DESCRIPTION
**Pull request by raul-te**
This pull request is the first of a set of refactories with breaking changes. The objective is mitigate a class of bugs ([for](https://github.com/thousandeyes/terraform-provider-thousandeyes/issues/59) [example](https://github.com/thousandeyes/terraform-provider-thousandeyes/issues/58)) where the JSON marshaling fails to distinguish between unset fields and those set to a zero-value. One solution is to use pointers for struct fields. This concept isn't novel, I've taken it from the [go-github wrapper](https://github.com/google/go-github/blob/master/README.md?plain=1#L201-L217).

Additionally, I'm including two other changes:
- `TestID` is now int64. In general, I want to future-proof our IDs to support > 2^32. I'll prepare another pull request modifying the rest of the ID fields. 
- I'm dropping `omitempty` for `alertRules`. Failing to set this field makes ThousandEyes assign default alert rules, which confuses the Terraform state. Better to be explicit and assign alert rules via Terraform.

All tests are passing after these changes:
```
$ go test
2022/05/20 16:12:46 [INFO] 100 of 240 requests / min remain.  Sleeping 750ms to prevent rate limiting.
2022/05/20 16:12:46 [INFO] 100 of 240 requests / min remain.  Sleeping 2s to prevent rate limiting.
2022/05/20 16:12:46 [INFO] 100 of 240 requests / min remain.  Sleeping 2s to prevent rate limiting.
2022/05/20 16:12:46 [INFO] 100 of 240 requests / min remain.  Sleeping 250ms to prevent rate limiting.
2022/05/20 16:12:46 [INFO] Rate Limited: Sleeping 31s before resubmitting
2022/05/20 16:12:46 [INFO] Rate Limited: Sleeping 31s before resubmitting
2022/05/20 16:12:46 [INFO] 240 of 240 requests / min remain.  Sleeping 250ms to prevent rate limiting.
2022/05/20 16:12:46 [INFO] Rate Limited: Sleeping 1m0s before resubmitting
PASS
ok  	github.com/thousandeyes/go-thousandeyes	0.100s

```

### CODE REVIEW

File: utils.go [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package thousandeyes
+
+// Bool is a helper routine that allocates a new bool value
+// to store v and returns a pointer to it.
+func Bool(v bool) *bool { return &v }
```
Comment (sfreitas-te) [inline utils.go:5]: Is this a common practice in Go?

**Review by sfreitas-te — APPROVED** [review]
Overall looks good left a couple of questions

File: utils.go (hunk 1 above)
Comment (raul-te) [reply utils.go:5]: I'm not sure if it's common, but I think it makes sense for our use-case. I took it verbatim from [Google's go-github API wrapper library.](https://github.com/google/go-github/blob/7a7a7f0/github/github.go#L1352-L1366)

**Review by xvilaca-te — DISMISSED** [review]
Some minor comments.

File: http_server.go [hunk 2]
```diff
@@ -20,62 +20,61 @@ type CustomHeaders struct {
...
+	ModifiedBy         *string             `json:"modifiedBy,omitempty"`
+	ModifiedDate       *string             `json:"modifiedDate,omitempty"`
+	SavedEvent         *int                `json:"savedEvent,omitempty"`
 	SharedWithAccounts []SharedWithAccount `json:"sharedWithAccounts,omitempty"`
-	TestID             int                 `json:"testId,omitempty"`
-	TestName           string              `json:"testName,omitempty"`
-	Type               string              `json:"type,omitempty"`
-	// LiveShare is common to all tests except DNS+
-	LiveShare int `json:"liveShare,omitempty"`
+	TestID             *int64              `json:"testId,omitempty,omitempty"`
```
[the remark this answers is not in the document]
Comment (raul-te) [reply http_server.go]: Oops, good catch. Fixed.

File: alerts.go [hunk 3]
```diff
@@ -31,33 +31,33 @@ type AlertRules []AlertRule
...
-  Message   string   `json:"message,omitempty"`
-  Recipient []string `json:"recipient,omitempty"`
+	Message   *string  `json:"message,omitempty"`
+	Recipient []string `json:"recipient,omitempty"`
 }
 
 // Notification - Alert Rule Notification structure
 type Notification struct {
-  Email NotificationEmail `json:"email,omitempty"`
+	Email NotificationEmail `json:"email,omitempty"`
```
[the remark this answers is not in the document]
Comment (raul-te) [reply alerts.go:40]: This is a good point. On this contribution I focused on `string` and `int` types. Based on some tests I did, it seems we can avoid doing "pointerizing" `slice` types. That leave us with struct types. I believe we'll need to make the change on those as well. I'll do that on a new pull request.

**Comment by raul-te** [comment]
I'll merge and apply  the recommendations on a new pull request. This one is already too bloated.

**Pull request merged**